src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c

```diff
@@ -0,0 +1,806 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifdef HEADLESS
+#error This file should not be included in headless library
+#endif
+
+#include <dlfcn.h>
+#include "jni_util.h"
+#include "awt_p.h"
+#include "awt.h"
+#include "screencast_pipewire.h"
+#include "fp_pipewire.h"
+#include <stdio.h>
+
+#include "gtk_interface.h"
+#include "gtk3_interface.h"
+
+int DEBUG_SCREENCAST_ENABLED = FALSE;
+
+struct ScreenSpace screenSpace = {0};
+static struct PwLoopData pw = {0};
+
+extern JavaVM *jvm;
+jclass screencastHelperClass = NULL;
+jmethodID storeTokenMethodID = NULL;
+
+inline void debug_screencast(
+        const char *__restrict fmt,
+        ...
+) {
+    if (DEBUG_SCREENCAST_ENABLED) {
+        va_list myargs;
+        va_start(myargs, fmt);
+        vfprintf(stdout, fmt, myargs);
+        va_end(myargs);
+    }
+}
+
+/**
+ * @return TRUE on success
+ */
+static gboolean initScreenSpace() {
+    screenSpace.screenCount = 0;
+    screenSpace.allocated = SCREEN_SPACE_DEFAULT_ALLOCATED;
+    screenSpace.screens = calloc(
+            SCREEN_SPACE_DEFAULT_ALLOCATED,
+            sizeof(struct ScreenProps)
+    );
+
+    if (!screenSpace.screens) {
+        ERR("failed to allocate memory\n");
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static void doCleanup() {
+    for (int i = 0; i < screenSpace.screenCount; ++i) {
+        struct ScreenProps *screenProps = &screenSpace.screens[i];
+        if (screenProps->data) {
+            if (screenProps->data->stream) {
+                fp_pw_stream_disconnect(screenProps->data->stream);
+                fp_pw_stream_destroy(screenProps->data->stream);
+                screenProps->data->stream = NULL;
+            }
+            free(screenProps->data);
+            screenProps->data = NULL;
+        }
+    }
+
+    if (pw.pwFd > 0) {
+        close(pw.pwFd);
+        pw.pwFd = -1;
+    }
+
+    portalScreenCastCleanup();
+
+    if (pw.core) {
+        fp_pw_core_disconnect(pw.core);
+        pw.core = NULL;
+    }
+
+    DEBUG_SCREENCAST("STOPPING loop\n", NULL)
+
+    if (pw.loop) {
+        fp_pw_thread_loop_stop(pw.loop);
+        fp_pw_thread_loop_destroy(pw.loop);
+        pw.loop = NULL;
+    }
+
+    if (screenSpace.screens) {
+        free(screenSpace.screens);
+        screenSpace.screens = NULL;
+    }
+}
+
+/**
+ * @return TRUE on success
+ */
+static gboolean initScreencast(const gchar *token) {
+    fp_pw_init(NULL, NULL);
+
+    pw.pwFd = -1;
+
+    if (!initScreenSpace()
+        || !initXdgDesktopPortal()
+        || (pw.pwFd = getPipewireFd(token)) < 0) {
+        doCleanup();
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+static inline void convertRGBxToBGRx(int* in) {
```

Comment:
Will this work correctly on little-endian ARM systems?

Reply:
I haven't tested it, but I'll try to find a machine to test it on.